[5.7] [WIP] Replace Fluent in Schema builder for columns

aleksandrlat opened the pull request:
_This is early preview pull request to show main idea and get feedback. If it looks good I can finish it._

The main idea of changes is to improve Schema builder with explicitly defined classes for column building instead of usage of Fluent. So code is easier to understand because new classes explicitly define possible functionality.

Also IDE can autocomplete possible methods:
![autocomplete1](https://user-images.githubusercontent.com/3708309/35478006-28818938-03e3-11e8-8ee1-ef2533fb72c0.png)

For example text specific methods like charset and collation will be autocompleted only for text column types:
![autocomplete2](https://user-images.githubusercontent.com/3708309/35478013-5e7b57a8-03e3-11e8-93b3-afe659a39c85.png)

For example for integer column type IDE can warn that method does not exist:
![autocomplete3](https://user-images.githubusercontent.com/3708309/35478019-7a01ea00-03e3-11e8-85ba-da3a757bb60a.png)

Because of usage of Fluent for now it is impossible to get such static analyze of code to predict some errors.

aleksandrlat committed:
- create classes for different column types instead of Fluent container usage
- use new classes

aleksandrlat committed:
- add doc comments

aleksandrlat committed:
- remove wrong doc blocks

GrahamCampbell commented on src/Illuminate/Database/Schema/Blueprint.php:
```diff
@@ -24,7 +24,7 @@ class Blueprint
      * The columns that should be added to the table.
      *
-     * @var \Illuminate\Support\Fluent[]
+     * @var Columns\Column[]
```
Please use fully qualified names in all phpdoc.

aleksandrlat replied:
Done

carusogabriel commented on src/Illuminate/Database/Schema/Columns/CharString.php:
```diff
@@ -0,0 +1,30 @@
+     * @param string $name
+     * @param int $length
+     */
+    public function __construct(string $type, string $name, int $length)
```
Please, don't use PHP 7 type hints, is hard to fix conflicts :+1:

crynobone replied:
How is it hard when this is a brand new class, what's it going to conflict with?  😂

ntzm replied:
Might as well follow the rest of the framework code tbh

aleksandrlat replied:
I see PHP 7 is requirement now. Should scalar types hints be used? I think this is great feature and helps a lot. It is impossible to add them in whole code base but they are still useful.
If it is requirement to avoid usage of scalar type hints I will remove them

aleksandrlat replied:
@carusogabriel could you please help me to understand why it is hard to fix conflicts in case of scalar type hints are used?

mfn commented on src/Illuminate/Database/Schema/Columns/Column.php:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    /**
+     * Add a comment to a column (MySQL)
```
These database specific comments are problematic, for example: postgres supports comments on columns too.

aleksandrlat replied:
I copied these comments from docs https://github.com/laravel/docs/blob/master/migrations.md#column-modifiers
Postgres supports comments but laravel schema build does not support them for postgres and for other DBs.

mfn commented:
@aleksandrlat that's quite some work tone here, kudos!

It improves the situation that the type hints are more correct, but my understanding is that `Column` is still database-un-specific and if the respective driver doesn't implement a feature, you would still think it could work when in fact does't? Because this is also one of the complains with `Fluent`, it's so easy to "forget" certain parameters.

aleksandrlat committed:
- use absolute class names in doc comments
- rename CharString to VariableLength
- fix issue with applying modifiers to columns
- create Illuminate\Database\Schema\Grammars\Grammar::isColumnSerial helper method and use in modifyIncrement methods

ntzm commented on src/Illuminate/Database/Schema/Columns/Decimal.php:
```diff
@@ -0,0 +1,67 @@
+ * Class Decimal
+ *
+ * @property-read int|null $total
+ * @property-read int|null $places
```
Any reason why you have a mix of `foo|null` and `?foo`?

aleksandrlat replied:
As I know these php docs are part of [PhpDocumentor](http://docs.phpdoc.org/guides/types.html#primitives). It does not support nullable types. I could be mistaken.

aleksandrlat replied:
Because of copy/paste:) Thank you!
I deleted ?type from everywhere because it is actually not required

aleksandrlat commented:
@mfn yes you are right. I was thinking about this too. But I just want to start with current changes only. 
As second step I can work on DB vendor specific columns and Blueprints.

aleksandrlat committed:
- delete default values for bool properties of Column to make behaviour more backward compatible with Fluent
- fix collation visibility issue
- fix modifyUnsigned
- fix test because useCurrent can be used only with timestamps

aleksandrlat committed:
- delete '?' from ?type because it is not actually required

taylorotwell commented:
I doubt I will be merging this.

ntzm commented:
Very sad to see this closed with little explanation, especially with 10 👍  votes

aleksandrlat commented:
> I doubt I will be merging this.

@taylorotwell why not? What is wrong with this? Any input would be appreciated.

deleugpn commented:
I would take a guess and say it's a huge core change that will only contribute with IDE auto-complete. [Source](https://github.com/laravel/framework/pull/22632#issuecomment-355556232).
I'm not saying auto complete is a small achievement (and I think it would be pretty neat to have it), but it does mean changing a lot of the core files without changing any code-execution whatsoever.
